flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -339,78 +343,107 @@ public void checkpointState(
+                "Notification of aborted checkpoint {} for task {}",
+                null,
+                (opw) -> opw.notifyCheckpointComplete(checkpointId),
+                (tsm) -> tsm.notifyCheckpointComplete(checkpointId));
     }
 
     @Override
     public void notifyCheckpointAborted(
             long checkpointId, OperatorChain<?, ?> operatorChain, Supplier<Boolean> isRunning)
             throws Exception {
 
+        notifyCheckpoint(
+                checkpointId,
+                operatorChain,
+                isRunning,
+                "Ignoring notification of aborted checkpoint {} for not-running task {}",
+                "Ignoring notification of aborted checkpoint {} for finished on restore task {}",
+                "Notification of aborted checkpoint {} for task {}",
+                () -> {
+                    boolean canceled = cancelAsyncCheckpointRunnable(checkpointId);
+
+                    if (!canceled) {
+                        if (checkpointId > lastCheckpointId) {
+                            // only record checkpoints that have not triggered on task side.
+                            abortedCheckpointIds.add(checkpointId);
+                        }
+                    }
+
+                    channelStateWriter.abort(
+                            checkpointId,
+                            new CancellationException("checkpoint aborted via notification"),
+                            false);
+                },
+                (opw) -> opw.getStreamOperator().notifyCheckpointAborted(checkpointId),
+                (tsm) -> tsm.notifyCheckpointAborted(checkpointId));
+    }
+
+    @Override
+    public void notifyCheckpointSubsumed(
+            long checkpointId, OperatorChain<?, ?> operatorChain, Supplier<Boolean> isRunning)
+            throws Exception {
+        notifyCheckpoint(
+                checkpointId,
+                operatorChain,
+                isRunning,
+                "Ignoring notification of subsumed checkpoint {} for not-running task {}",
+                "Ignoring notification of subsumed checkpoint {} for finished on restore task {}",
+                "Notification of subsumed checkpoint {} for task {}",
+                null,
+                (opw) -> opw.notifyCheckpointSubsumed(checkpointId),
+                null);
+    }
+
+    private void notifyCheckpoint(
+            long checkpointId,
+            OperatorChain<?, ?> operatorChain,
+            Supplier<Boolean> isRunning,
+            String logFormatIfNotRunning,
+            String logFormatIfFinishedOnRestore,
+            String logFormatIfNotificationStart,
+            @Nullable Runnable runnableBeforeNotifyOperator,
+            @Nonnull ThrowingConsumer<StreamOperatorWrapper<?, ?>, Exception> operatorConsumer,
+            @Nullable ThrowingConsumer<TaskStateManager, Exception> taskStateManagerConsumer)
+            throws Exception {
```

Comment:
Instead of using `Runnable` and `ThrowingConsumer` in order to deduplicate the code, I would suggest to consider deduplicating it by decomposing this `notifyCheckpoint` method into smaller parts. For example:

```

public void notifyCheckpointAborted(...) {
  if (notifyCheckpointOperationrShouldBeExecuted()) {
    // inlined content of the `runnableBeforeNotifyOperator`
    
    operatorChain.forEach((operatorWrapper) -> operatorWrapper.notifyCheckpointSubsumed(checkpointId));
    
    env.getTaskStateManager().notifyCheckpointAborted(checkpointId);
  }
}

public void notifyCheckpointSubsumed(...) {
  if (notifyCheckpointOperationrShouldBeExecuted()) {
    operatorChain.forEach((operatorWrapper) -> operatorWrapper.notifyCheckpointAborted(checkpointId));
  }
}

private boolean notifyCheckpointOperationrShouldBeExecuted(...) {
  // isRunning/isFinishedOnRestore checks + logging
}
```

Yes, we would still have a bit of code duplication, but it would avoid the problem of generic `Runnable` `ThrowingConsumer` that are mangling/hiding what the code is doing. In your current version to understand what this code is doing, you have to keep jumping back and forth between `notifyCheckpointAborted()` method (where those `Runnable` and `ThrowingConsumers` are defined) and the actual `notifyCheckpoint` implementation. What makes it even worse, is that you can not use "find implementations" from the IDE to help you, as there are just so many `Runnable`s in the project. So maybe a bit of code duplication is not that bad in that case? (I'm not 100% sure)

Reply:
The refactor also take the `enum` of `NotifyCheckpointOperation` as `Task` did.